### PR TITLE
Feat/sqlite jsonb

### PR DESCRIPTION
**Pull request by siddhantk232**
This works and I can compile the sample-sqlite wasm app. But, I think we should not be using the `SqliteType::Binary`. I haven't tested it yet on a real database.

### CODE REVIEW

**Comment by siddhantk232** [comment]
After testing on a real sqlite db, I get the following output:

```
wasm: hello wasm: /
sqlite_connect: default
wasm: load
wasm: q: Query { sql: "SELECT `json_data`.`data` FROM `json_data`", binds: [] }
q: {"sql":"SELECT `json_data`.`data` FROM `json_data`","binds":[]}
conn, sql: SELECT `json_data`.`data` FROM `json_data`
stmt
wasm: json_from_ptr
wasm: json_from_ptr: {"Ok":{"columns":["data"],"rows":[{"fields":[{"Text":"{\"id\": 1}"}]},{"fields":[{"Integer":0}]}]}}
wasm: json_data table res: Err(DeserializationError("Unexpected type"))
wasm: load
wasm: q: Query { sql: "INSERT INTO `json_data` (`id`, `data`) VALUES (?, ?) RETURNING `json_data`.`id`", binds: [Integer(69), Blob([123, 34, 97, 103, 101, 34, 58, 50, 57, 44, 34, 99, 111, 110, 102, 105, 103, 34, 58, 123, 34, 112, 114, 101, 102, 101, 114, 115, 95, 101, 109, 97, 105, 108, 34, 58, 116, 114, 117, 101, 125, 44, 34, 110, 97, 109, 101, 34, 58, 34, 74, 111, 104, 110, 34, 44, 34, 112, 104, 111, 110, 101, 115, 34, 58, 91, 34, 43, 52, 52, 32, 49, 50, 51, 52, 53, 54, 55, 34, 44, 34, 43, 52, 52, 32, 50, 51, 52, 53, 54, 55, 56, 34, 93, 125])] }
q: {"sql":"INSERT INTO `json_data` (`id`, `data`) VALUES (?, ?) RETURNING `json_data`.`id`","binds":[{"Integer":69},{"Blob":[123,34,97,103,101,34,58,50,57,44,34,99,111,110,102,105,103,34,58,123,34,112,114,101,102,101,114,115,95,101,109,97,105,108,34,58,116,114,117,101,125,44,34,110,97,109,101,34,58,34,74,111,104,110,34,44,34,112,104,111,110,101,115,34,58,91,34,43,52,52,32,49,50,51,52,53,54,55,34,44,34,43,52,52,32,50,51,52,53,54,55,56,34,93,125]}]}
conn, sql: INSERT INTO `json_data` (`id`, `data`) VALUES (?, ?) RETURNING `json_data`.`id`
stmt
wasm: json_from_ptr
wasm: json_from_ptr: {"Ok":{"columns":["id"],"rows":[{"fields":[{"Integer":69}]}]}}
wasm: affected: Ok(69)
```

the insert seem to have worked but the select throws `DeserializationError`

cc: @amitu

**Comment by amitu** [comment]
> But, I think we should not be using the SqliteType::Binary.

This is what I understand, `jsonb` support means there is a binary encoding, which sqlite uses to store the data. The data is still stored in BLOB data type. We could have use any encoding, and do not have to worry about their encoding, but they come with a lot of json related functions and query syntax. These functions used to work only with JSON stored as TEXT, but now they work with either JSON in TEXT or JSONB in BLOB (these functions auto detect based on data, and within same table, different rows can have either JSON in TEXT or JSONB in BLOB formats, and queries on this table will just work). 

Now the question is how to convert data to JSONB binary encoding? One option is we convert it to JSON as text, and pass to a function `jsonb(json_as_str)`. This is what you would do if you were directly using CLI:  `insert ... ($id, jsonb('{"some": 1}'))`.

Now for us the question is if we somehow use this `jsonb()` function, or directly from Rust convert to their encoding, there is a crate for that: [serde-sqlite-jsonb](https://github.com/lovasoa/serde-sqlite-jsonb). Normally I would have resisted as this is binary encoding of another software, and what if it changes in future, but their encoding is [well documented](https://sqlite.org/draft/jsonb.html):

> However, JSONB is intended to be portable and backwards compatible for all future versions of SQLite. In other words, you should not have to export and reimport your SQLite database files when you upgrade to a newer SQLite version. For that reason, the JSONB format needs to be well-defined.
>
> This document is therefore similar in purpose to the [SQLite database file format](https://sqlite.org/draft/fileformat2.html) document that describes the on-disk format of an SQLite database file. Applications are not expected to directly read and write the bits and bytes of SQLite database files. The SQLite database file format is carefully documented so that it can be stable and enduring. In the same way, the JSONB representation of JSON is documented here so that it too can be **stable and enduring**, not so that applications can read or writes individual bytes.

Since they are committed to "stable and enduring" encoding format, it may be okay for us to rely on this. Other alternative is we convert to text and then json jsonb function somehow (I am not even sure how to do that with diesel's connection support, and even if we could it would be double encoding and thus wasteful).

File: ft-sys/src/diesel_sqlite/sqlite_value.rs [hunk 1]
```diff
@@ -37,6 +37,13 @@ impl<'a> SqliteValue<'a> {
             _ => Err("Unexpected type".into()),
         }
     }
+
+    pub(crate) fn jsonb(&self) -> diesel::deserialize::Result<serde_json::Value> {
+        match self.raw_value {
+            Value::Jsonb(i) => Ok(serde_json::from_slice(i)?),
+            _ => Err("Unexpected type".into()),
```
Comment (amitu) [inline ft-sys/src/diesel_sqlite/sqlite_value.rs:44]: Update these error messages to include the "got type" part also, so it's easier to debug.

**Comment by amitu** [comment]
> `wasm: json_from_ptr: {"Ok":{"columns":["data"],"rows":[{"fields":[{"Text":"{\"id\": 1}"}]},{"fields":[{"Integer":0}]}]}}`
> `wasm: json_data table res: Err(DeserializationError("Unexpected type"))`

@siddhantk232 I think this is because SQLite is treating it as Text. If you check `sqlite_value.rs`, all the "deserializers" do a type check, and only return the data if correct type. I think this is happening because SQLite is mistakenly assuming our data is Text because it is full of ascii characters (I am not very confident about this hypothesis).

File: ft-sys/src/diesel_sqlite/sqlite_value.rs [hunk 2]
```diff
@@ -46,6 +53,7 @@ pub enum Value {
     Real(f64),
     Text(String),
     Blob(Vec<u8>),
+    Jsonb(Vec<u8>),
```
Comment (amitu) [inline ft-sys/src/diesel_sqlite/sqlite_value.rs:56]: I think we should remove this clause and use Blob only.

**Comment by amitu** [comment]
@siddhantk232 Are you planning to use json syntax/functions in your queries?

**Comment by siddhantk232** [comment]
> @siddhantk232 Are you planning to use json syntax/functions in your queries?

Nope

**Comment by amitu** [comment]
I still recommend we implement this PR with the rust crate doing serialisation and we using blob internally, so we can use it and it's def a good feature for our customers.

File: ft-sys/src/diesel_sqlite/types/mod.rs [hunk 3]
```diff
@@ -125,6 +125,20 @@ impl ToSql<sql_types::Double, Sqlite> for f64 {
...
 
+impl ToSql<sql_types::Jsonb, Sqlite> for serde_json::Value {
+    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Sqlite>) -> serialize::Result {
+        let b = serde_json::to_vec(self)?;
+        out.set_value(b);
+        Ok(IsNull::No)
+    }
+}
+
+impl FromSql<sql_types::Jsonb, Sqlite> for serde_json::Value {
```
Comment (amitu) [inline ft-sys/src/diesel_sqlite/types/mod.rs:136]: Ideally we should implement `FromSql<sql_types::Jsonb, Sqlite>` and `ToSql<sql_types::Jsonb, Sqlite>` for any type `T` that implements `Serialize`/`Deserialize` traits.

**Review by amitu — APPROVED** [review]

**Pull request merged**